libraries/persistent_store/src/lib.rs

```diff
@@ -12,10 +12,336 @@
+//!
+//! By induction, we have:
+//!
+//! ```text
+//! t_{I + i} <= t_I + |h_{I + i}| - |h_I| + i
+//! ```
+//!
+//! We have the following properties:
+//!
+//! ```text
+//! t_I <= h_I + V
+//! |h_{I + i}| - |h_I| <= h_{I + i} - h_I
+//! h_{I + i} <= (I + i) * Q + M
+//! ```
+//!
+//! Replacing into our previous equality, we can conclude:
+//!
+//! ```text
+//! t_{I + i}  = t_I + |h_{I + i}| - |h_I| + i
+//!           <= h_I + V + (I + i) * Q + M - h_I + i
+//!            = (N - 1) * (Q - 1) - M + (I + i) * Q + M + i
+//!            = (N - 1) * (Q - 1) + (I + i) * Q + i
+//!            = (I + i + N - 1) * Q + i - (N - 1)
+//!           <= (I + i + N - 1) * Q
+//! ```
+//!
+//! We also want to show that after `N - 1` compactions, the remaining capacity is
+//! available without compaction.
+//!
+//! ```text
+//! V - (t_{I + N - 1} - h_{I + N - 1}) >=    // The available words in the window.
+//!   C - (|t_{I + N - 1}| - |h_{I + N - 1}|) // The remaining capacity.
+//!   + 1                                     // Reserved for Clear.
+//! ```
+//!
+//! We can replace the definition of `C` and simplify:
+//!
+//! ```text
+//! V - (t_{I + N - 1} - h_{I + N - 1}) >= V - N - (|t_{I + N - 1}| - |h_{I + N - 1}|) + 1
+//! iff t_{I + N - 1} - h_{I + N - 1} <= |t_{I + N - 1}| - |h_{I + N - 1}| + N - 1
+//! ```
+//!
+//! We have the following properties:
+//!
+//! ```text
+//! t_{I + N - 1} = t_I + |h_{I + N - 1}| - |h_I| + N - 1
+//! |t_{I + N - 1}| - |h_{I + N - 1}| = |t_I| - |h_I| // Compaction preserves capacity.
+//! |h_{I + N - 1}| - |t_I| <= h_{I + N - 1} - t_I
+//! ```
+//!
+//! From which we conclude:
+//!
+//! ```text
+//! t_{I + N - 1} - h_{I + N - 1} <= |t_{I + N - 1}| - |h_{I + N - 1}| + N - 1
+//! iff t_I + |h_{I + N - 1}| - |h_I| + N - 1 - h_{I + N - 1} <= |t_I| - |h_I| + N - 1
+//! iff t_I + |h_{I + N - 1}| - h_{I + N - 1} <= |t_I|
+//! iff |h_{I + N - 1}| - |t_I| <= h_{I + N - 1} - t_I
+//! ```
+//!
+//!
+//! ## Checksum
+//!
+//! The main property we want is that all partially written/erased words are either
+//! the initial word, the final word, or invalid.
+//!
+//! We say that a bit sequence `TARGET` is reachable from a bit sequence `SOURCE` if
+//! both have the same length and `SOURCE & TARGET == TARGET` where `&` is the
+//! bitwise AND operation on bit sequences of that length. In other words, when
+//! `SOURCE` has a bit equal to 0 then `TARGET` also has that bit equal to 0.
+//!
+//! The only written entries start with `101` or `110` and are written from an
+//! erased word. Marking an entry as padding or deleted is a single bit operation,
+//! so the property trivially holds. For those cases, the proof relies on the fact
+//! that there is exactly one bit equal to 0 in the 3 first bits. Either the 3 first
+//! bits are still `111` in which case we expect the remaining bits to bit equal
+//! to 1. Otherwise we can use the checksum of the given type of entry because those
+//! 2 types of entries are not reachable from each other.
```

Comment:
To me, this lacks some introduction to what the "checksum of the given type" actually is. It seems to prove something without explaining the actual point of a checksum, or what it is used for.

Reply:
The first sentence of this checksum section was supposed to define what checksums are for. I could clarify if I understand what's missing.

What I mean by "checksum of the given type" is that entries can be of different types (user entry or internal entry). Each of those has a checksum field (last 6 bits for user and 5 for internal). This is what I mean by the checksum of the given type. So to try to reformulate the idea of those last 2 sentences: If a word starts with `111` then it has to be fully erased to be valid (and is not an entry). Otherwise, it has to be an entry, so we check the checksum of that type of entry. We find the type of the entry by looking at the initial bits: if it's `101` it's a user entry, if it's `110` it's an internal entry.

Reply:
It is guessable that different types have different checksums, and the checksum type to be expected is given through the header bits.

This section is like a Maths lecture. You know the definition, and you can verify that everything you are saying is true. I was just wondering if this is what you want in documentation. Maybe add a summary of the differences between checksums, like with a ASCII art visualisation?

Reply:
Yes, this is what this sub-section is about. It is part of the Proofs section which is not proper documentation but artifacts to help speed up a manual verification of the properties. Because the chain of thoughts is not always trivial and having a written trace is useful.

Good point for better visualisation. Added.